pkg/v1/cache/fs.go

```diff
@@ -0,0 +1,80 @@
+package cache
+
+import (
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/tarball"
+)
+
+type fscache struct {
+	path string
+}
+
+// NewFilesystemCache returns a Cache implementation backed by files.
+func NewFilesystemCache(path string) Cache {
+	return &fscache{path}
+}
+
+func (fs *fscache) Put(l v1.Layer) (v1.Layer, error) {
+	h, err := l.Digest()
+	if err != nil {
+		return nil, err
+	}
+	path := filepath.Join(fs.path, h.String())
+	return &layer{Layer: l, path: path}, nil
+}
+
+type layer struct {
+	v1.Layer
+	path string
+}
+
+func (l *layer) Compressed() (io.ReadCloser, error) {
+	if err := os.MkdirAll(filepath.Dir(l.path), 0700); err != nil {
+		return nil, err
+	}
+	f, err := os.Create(l.path)
+	if err != nil {
+		return nil, err
+	}
+	rc, err := l.Layer.Compressed()
+	if err != nil {
+		return nil, err
+	}
+	t := io.TeeReader(rc, f)
+	closes := []func() error{rc.Close, f.Close}
+	return &readcloser{t: t, closes: closes}, nil
+}
+
+type readcloser struct {
+	t      io.Reader
+	closes []func() error
+}
+
+func (rc *readcloser) Read(b []byte) (int, error) {
+	return rc.t.Read(b)
+}
+
+func (rc *readcloser) Close() error {
+	for _, c := range rc.closes {
+		if err := c(); err != nil {
+			return err
+		}
```

Comment:
Sorry for the drive-by review -- you might want to make a best effort to always call every close function, even if one of them returns an error. Most clients don't handle `Close()` errors (e.g. `defer x.Close()`), so if the first close fails, no one will notice that the second close (i.e. closing the file) was never called.

Reply:
Excellent point, thanks for the drive-by review! (I'm not sure it's a drive-by if I @ you on it though 😆 )